2.5-huffman.ipynb

```diff
@@ -1,25 +1,207 @@
 {
  "cells": [
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import util\n",
+    "from test_framework.key import ECKey, ECPubKey\n",
+    "from test_framework.script import TapTree, TapLeaf, Node"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# 2.5 Huffman TapTree Constructor\n",
+    "\n",
+    "[Huffman's algorithm](https://en.wikipedia.org/wiki/Huffman_coding) can optimize the TapTree structure, and thus potentially lower the size of revealed branches, by taking into account the estimated frequency by which each TapLeaf will occur. Since some spending scenarios are more likely to occur than others, we would ideally optimize the tree structure such that those spends appear closer to the root of the tree, and therefore require a smaller merkle proof when spending."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "To construct a TapTree with the Huffman algorithm, we simply assign a relative probability weight to each leaf. \n",
+    "\n",
+    "A higher `leaf weight` represents:\n",
+    "\n",
+    "* A higher likelihood of execution\n",
+    "* A position which is closer to the root (if possible)\n",
+    "* A smaller inclusion proof"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Part 1: Huffman Algorithm\n",
+    "\n",
+    "The Huffman algorithm to construct a TapTree structure can be described as follows:\n",
+    "\n",
+    "* Place all leafs in a node queue and sort by weight*\n",
+    "* While length ( node queue ) > 1 :\n",
+    "    * Pop the two elements with the highest weights\n",
+    "    * Generate parent node from popped elements\n",
+    "    * Add parent node to queue and resort\n",
+    "* Root node is represented by final queue element"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/huffman_intro0.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    ">\\* If more than 2 leafs or nodes share the same weight during construction with the Huffman algorithm, the selection of the two nodes to generate the next parent becomes ambiguous. In our implementation, we first sort our node queue by `weight`, then by `tagged hash` value. This way, the sorting of the node queue is always unambiguous given a distinct set of leafs."
```

Comment:
This is rendered as a quote. Is that what you intended?

Reply:
Yes, but I doesn't look right, changed it to normal paragraph.